### ISSUE TITLE
Exception while reading data: TypeError: 'float' object is not iterable

### ISSUE THREAD

**Issue by afilipchik** [graphite-project/graphite-web#1156]
Deployed latest graphite-web yesterday and now see a lot of:

Traceback (most recent call last):
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/opt/graphite/webapp/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/opt/graphite/webapp/graphite/readers.py", line 174, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable

**Comment by afilipchik** [comment]
the hacky fix is to add:
 if hasattr(cached_datapoints, 'iteritems'):
        cached_datapoints = cached_datapoints.iteritems()


try:
      cached_datapoints = CarbonLink.query(self.real_metric_path)
    except:
      log.exception("Failed CarbonLink query '%s'" % self.real_metric_path)
      cached_datapoints = {}

      if hasattr(cached_datapoints, 'iteritems'):
        cached_datapoints = cached_datapoints.iteritems()

      for (timestamp, value) in cached_datapoints:
        interval = timestamp - (timestamp % data.timeStep)

        try:
          i = int(interval - data.startTime) / data.timeStep
          values[i] = value
        except:
          pass

**Comment by loopping** [comment]
I tried your hacky fix with no luck:

<pre><code>webapp/cache.log
Sat Feb 14 14:53:27 2015 :: CarbonLink sending request for carbon.agents.vmlinux-a.cache.size to ('127.0.0.1', 'a')
Sat Feb 14 14:53:27 2015 :: CarbonLink finished receiving carbon.agents.vmlinux-a.cache.size from ('127.0.0.1', 'a')
Sat Feb 14 14:53:27 2015 :: CarbonLink sending request for carbon.agents.vmlinux-a.cache.size to ('127.0.0.1', 'b')
Sat Feb 14 14:53:27 2015 :: CarbonLink finished receiving carbon.agents.vmlinux-a.cache.size from ('127.0.0.1', 'b')
Sat Feb 14 14:53:27 2015 :: CarbonLink cache-query request for carbon.agents.vmlinux-a.cache.size returned 2 datapoints
</code></pre>
<pre><code>webapp/exception.log
Sat Feb 14 14:53:27 2015 :: Failed after 2 retry! See: 'float' object is not iterable
Traceback (most recent call last):
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/opt/graphite/webapp/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/opt/graphite/webapp/graphite/readers.py", line 177, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable
Sat Feb 14 14:53:27 2015 :: Exception encountered in <POST http://vmlinux:8080/render>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/graphite/webapp/graphite/render/views.py", line 111, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 7, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 18, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 21, in evaluateTokens
    return fetchData(requestContext, tokens.pathExpression)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 141, in fetchData
    raise Exception("Failed after %i retry! See: %s" % (settings.MAX_FETCH_RETRIES, e))
Exception: Failed after 2 retry! See: 'float' object is not iterable</code></pre>

**Comment by afilipchik** [comment]
Did you change it in both places (WhisperReader and CeresReader)? It looks like it should be fixed by:

https://github.com/graphite-project/graphite-web/pull/1122 but only for WhisperReader :(

**Comment by AlexAkulov** [comment]
I get the same error.
Эта ошибка появляется если я использую подключение через carbon link, и е очень часто когда Disk Utization 100%

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/django/core/handlers/base.py", line 115, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/opt/graphite-storage/webapp/graphite/render/views.py", line 111, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File "/opt/graphite-storage/webapp/graphite/render/evaluator.py", line 7, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File "/opt/graphite-storage/webapp/graphite/render/evaluator.py", line 18, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/opt/graphite-storage/webapp/graphite/render/evaluator.py", line 21, in evaluateTokens
    return fetchData(requestContext, tokens.pathExpression)
  File "/opt/graphite-storage/webapp/graphite/render/datalib.py", line 141, in fetchData
    raise Exception("Failed after %i retry! See: %s" % (settings.MAX_FETCH_RETRIES, e))
Exception: Failed after 2 retry! See: 'float' object is not iterable
```

**Comment by AlexAkulov** [comment]
I get the same error.
This error appears if i use the option CARBONLINK_HOSTS, and it doesn`t appear if i use only the option WHISPER_DIR in the file local_settings.py.
Also this error appears very often when Disk Utization reaches 100% (it doesn`t with WHISPER_DIR and hight Disk Utilization)

My configuration is the following
4 processes carbon-cache
1 gunicorn for graphite-wsgi

I`m ready to give any additional information. 

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/django/core/handlers/base.py", line 115, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/opt/graphite-storage/webapp/graphite/render/views.py", line 111, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File "/opt/graphite-storage/webapp/graphite/render/evaluator.py", line 7, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File "/opt/graphite-storage/webapp/graphite/render/evaluator.py", line 18, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/opt/graphite-storage/webapp/graphite/render/evaluator.py", line 21, in evaluateTokens
    return fetchData(requestContext, tokens.pathExpression)
  File "/opt/graphite-storage/webapp/graphite/render/datalib.py", line 141, in fetchData
    raise Exception("Failed after %i retry! See: %s" % (settings.MAX_FETCH_RETRIES, e))
Exception: Failed after 2 retry! See: 'float' object is not iterable
```

**Comment by obfuscurity** [comment]
This looks like a duplicate of #1139, which was resolved via #1122. If I'm incorrect please let me know and we can reopen.
